frontend_tests/node_tests/settings_user_groups.js

```diff
@@ -268,6 +289,163 @@ set_global('people', {
+        }
+        if (elem === '.description') {
+            user_group_find_called += 1;
+            return description_field_stub;
+        }
+    };
+
+    var pill_container_stub = $('.pill-container[data-group-pills="Mobile"]');
+    // Test the 'off' handlers on the pill-container
+
+    pill_container_stub.off = function (event_name, sel) {
+        if (sel === undefined) {
+           sel = 'whole';
+        }
+        turned_off[event_name + '/' + sel] = true;
+    };
+
+    pill_container_stub.on = function (event_name, handler) {
+        var event = {
+            stopPropagation: noop,
+        };
+        assert.equal(event_name, 'click');
+        on_click_called = true;
+        handler(event);
+    };
+
+    var pill_stub = $.create('fake-pill');
+
+    pill_container_stub.find = function (elem) {
+        if (elem === '.input') {
+            pill_container_find_called += 1;
+            return input_stub;
+        }
+        if (elem === '.pill') {
+            pill_container_find_called += 1;
+            return pill_stub;
+        }
+    };
+
+    pill_stub.hover = function (one, two) {
+        callback = one;
+        pill_hover_called = true;
+        assert.equal(typeof(one), 'function');
+        assert.equal(typeof(two), 'function');
+    };
+
+    var exit_button = $.create('fake-pill-exit');
+    pill_stub.set_find_results('.exit', exit_button);
+
+    exit_button.css = function (property, value) {
+        exit_button_called = true;
+        assert.equal(property, 'opacity');
+        assert.equal(value, '0.5');
+    };
+
+    // Test click handlers with an external user
+    $('#user-groups #1').on = function (event_name, child_selector, fn) {
+        assert.equal(fn(), undefined);
+    };
+
+    $('#user-groups').on = function (event_name, child_selector, fn) {
+        if (event_name === 'click' && child_selector === '.delete') {
+            var fake_delete = $.create('fk-#user-groups.delete_btn');
+            fake_delete.set_parents_result('.user-group', $('.user-group'));
+            $('.user-group').attr('id', '1');
+            assert.equal(fn.call(fake_delete), undefined);
+        }
+        if (event_name === 'keypress' && child_selector === '.user-group h4 > span') {
+            return noop;
+        }
+        if (event_name === 'input' && child_selector === '.user-group h4 > span') {
+            var fake_input = $.create('fk-#user-groups_input');
+            fake_input.set_parents_result('.user-group', $('.user-group'));
+            $('.user-group').attr('id', '2');
+            assert.equal(fn.call(fake_input), undefined);
+        }
+        if (event_name === 'click' && child_selector === '.save-group-changes') {
+            var fake_change = $.create('fk-#user-groups_click_save');
+            fake_change.set_parents_result('.user-group', $('.user-group'));
+            $('.user-group').attr('id', '3');
+            assert.equal(fn.call(fake_change), undefined);
```

Comment:
It seems like you should be making more assertions here about what the function actually does.

Reply:
@showell here I can't make any other assertions since main functionality is already being tested in test_on_events function, also since I am just returning from that click handler, so I can only assert that return value is undefined.

Reply:
If you're not making any meaningful assertions, then I would just remove the test code, since it's gonna give a false sense of assurance in the coverage reports.  Or make helper functions that you can call in both paths.